aten/src/ATen/native/Linear.cpp

```diff
@@ -53,20 +55,29 @@ static inline bool parseLinearFlatten3d() {
   return bool(value);
 }
 
-// `_flatten_3d_linear` flattens the first two dimensions of the input tensor
-// before passing it to linear operation, if the input tensor is 3D
-static inline Tensor _flatten_3d_linear(const Tensor& input, const Tensor& weight, const Tensor& bias) {
+// `_flatten_nd_linear` flattens all but the last dimension of the input tensor
+// before passing it to linear operation
+static inline Tensor _flatten_nd_linear(const Tensor& input, const Tensor& weight, const Tensor& bias) {
     const auto input_sizes = input.sym_sizes();
-    const auto result = at::addmm(bias, input.view_symint({input_sizes[0] * input_sizes[1], input_sizes[2]}), weight.t());
-    return result.view_symint({input_sizes[0], input_sizes[1], result.sym_size(1)});
+    // can't use -1 in reshape because it errors when a dimension is 0
+    c10::SymInt flattened_dim = 1;
+    for (size_t i = 0, ndim = input_sizes.size(); i < ndim - 1; ++i) {
```

Comment:
`ndim - 1` here hazardous because you are using unsigned `size_t` type, you will overflow if `ndim == 0`. Just use `int64_t`, the unsigned type here really is not worth it.